packages/cohort/src/replicator/services/replicator_service.rs

```diff
@@ -39,47 +53,91 @@ where
                     ChannelMessage::Decision(decision_version, decision_message) => {
                         replicator.process_decision_message(decision_version, decision_message).await;
 
-
+                        if total_items_processed == 0 {
+                            time_first_item_created_start_ns = OffsetDateTime::now_utc().unix_timestamp_nanos();
+                        }
                         // Get a batch of remaining versions with their statemaps to install.
-                        let statemaps_batch = replicator.generate_statemap_batch();
+                        let (all_versions_picked, statemaps_batch) = replicator.generate_statemap_batch();
+
+                        total_items_send += statemaps_batch.len();
+                        total_items_processed += all_versions_picked.len();
 
-                        if !statemaps_batch.is_empty() {
-                            statemaps_tx.send(statemaps_batch).await.unwrap();
+                        let statemap_batch_cloned = statemaps_batch.clone();
+                        let versions_not_send = all_versions_picked.into_iter().filter(|&v| {
+                            !statemap_batch_cloned.iter().any(|sm_b| sm_b.0 != v)
+                        });
+
+
+                        // Send statemaps batch to
+                        for (_, statemap_vec) in statemaps_batch {
+                            statemaps_tx.send(statemap_vec).await.unwrap();
                         }
 
+                        // These versions are decided but they are not send to Statemap installer as they are either aborted or don't have statemap
+                        versions_not_send.for_each(|version| {
+                            replicator.suffix.set_item_installed(version);
+
+                        });
+
+                        time_last_item_send_end_ns = OffsetDateTime::now_utc().unix_timestamp_nanos();
+
                     },
                 }
             }
         }
         // Commit offsets at interval.
         _ = interval.tick() => {
+            if config.enable_stats {
+                let duration_sec = Duration::from_nanos((time_last_item_send_end_ns - time_first_item_created_start_ns) as u64).as_secs_f32();
```

Comment:
Just sharing: I recently discovered more compact way to deal with time measurements:

```
let started_at = Instant::now();
....
let elapsed_ns = started_at.elapsed().as_nanos();
```

Reply:
nice, I will update this in the next PR where I will work on moving the stats as a struct as you have commented. Next PR will move replicator/installer into its own package and I will update this there.

Thanks for the suggestion :)